### ISSUE TITLE
babel.Transformer is not a function

### ISSUE THREAD

**Issue by mnzt** [sindresorhus/ava#207]
error
```
node: v4.2.2
npm: v1.4.9


C:\Users\mnzt\~code\ava-test\node_modules\ava\node_modules\babel-plugin-espower\create.js:6
    return new babel.Transformer('babel-plugin-espower', createEspowerVisitor(babel, options));
           ^

TypeError: babel.Transformer is not a function
    at createEspowerPlugin (C:\MEDIA\~code\Binaries\node_modules\ava\node_modules\babel-plugin-espower\create.js:6:12)
    at Object.<anonymous> (C:\MEDIA\~code\Binaries\node_modules\ava\lib\babel.js:21:3)
    at Module._compile (module.js:435:26)
    at Object.Module._extensions..js (module.js:442:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
    at Function.Module.runMain (module.js:467:10)
    at startup (node.js:136:18)
    at node.js:963:3
```
package.json
```
  "devDependencies": {
    "ava": "^0.4.2"
  },
```

**Comment by mnzt** [comment]
My bad, there was a stray babel package in `node_modules`

**Comment by scarletsky** [comment]
@mnzt 
I have the same issue, how can you solve it ?

**Comment by mnzt** [comment]
@scarletsky 

Go into your own `node_modules` and remove babel packages, then remove them from your `package.json`. After that, `npm install - -save-dev ava` should get it working. 

I think there may still be an issue with ava using local babel packages, I'll do some bug testing tonight and see if that's the case.

**Comment by scarletsky** [comment]
@mnzt 
Which babel version are you using? 5 or 6 ?

**Comment by mnzt** [comment]
@scarletsky I'm not running any babel packages in the project, due to those causing the error. 

Can you try removing your local babel packages in your project, and see if that temp-fixes things? If so, then I'll reopen this issue and we can provide extra debug info.

**Comment by scarletsky** [comment]
@mnzt 
after removing babel packages with `rm -rf node_modules/babel*`, I run tests again.

```js
// test.js
import test from 'ava'
test('foo', t => {
    t.pass();
    t.end()
});
```

```shell
$ ava test.js
```

```shell
Uncaught Exception:  /Users/scarlex/Projects/v1/test.js
  TypeError: (0 , _ava2.default) is not a function
    at Object.<anonymous> (/Users/scarlex/Projects/v1/test.js:9:22)

Uncaught Exception:  /Users/scarlex/Projects/v1/test.js
  TypeError: (0 , _ava2.default) is not a function
    at Object.<anonymous> (/Users/scarlex/Projects/v1/test.js:9:22)

Error: Never got test results from: /Users/scarlex/Projects/v1/test.js
    at ChildProcess.<anonymous> (/usr/local/lib/node_modules/ava/lib/fork.js:61:12)
    at emitTwo (events.js:87:13)
    at ChildProcess.emit (events.js:172:7)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
From previous event:
    at module.exports (/usr/local/lib/node_modules/ava/lib/fork.js:25:16)
    at run (/usr/local/lib/node_modules/ava/cli.js:120:9)
    at Array.map (native)
    at /usr/local/lib/node_modules/ava/cli.js:206:22
From previous event:
    at init (/usr/local/lib/node_modules/ava/cli.js:198:4)
    at Object.<anonymous> (/usr/local/lib/node_modules/ava/cli.js:243:2)
    at Module._compile (module.js:435:26)
    at Object.Module._extensions..js (module.js:442:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
    at Function.Module.runMain (module.js:467:10)
    at startup (node.js:136:18)
    at node.js:963:3
```

**Comment by mnzt** [comment]
@scarletsky well removing babel packages clears up the original issue, I'll put in a merge request later to fix if no one else does first. 

Unfortunately I can't debug your issue from mobile, hopefully some one else can help out in the mean time. It might be worth opening a separate issue for it.

**Comment by mnzt** [comment]
@sindresorhus was this fixed in `0.5.0`?

**Comment by scarletsky** [comment]
No, in fact, I am using 0.5.0...

**Comment by hellsan631** [comment]
@scarletsky I had this problem (TypeError: (0 , _ava2.default) is not a function) and was tinkering around trying to fix. Deleted my node_modules folder and accidentally ran ava and it worked. Apparently if you have a local copy of ava and a global copy of ava, it gives you the error you had above.

**Comment by mnzt** [comment]
@hellsan631 seems very variable, similar sort of issue with the original error with local modules interfering with ava's.

**Comment by mariuslundgard** [comment]
I get this error too. This package doesn't seem to be compatible with babel v6.

**Comment by mnzt** [comment]
Ava should really be using absolute paths for it's packages.

**Comment by jamestalmage** [comment]
We are not Babel v6 ready. #221 is in progress, but until some blocking issues get resolved - v6 is not an option. It will happen soon.

**Comment by jamestalmage** [comment]
```
  TypeError: (0 , _ava2.default) is not a function
```

This is a known issue in Babel v6: https://phabricator.babeljs.io/T6644

This is the primary blocker for AVA upgrading to Babel.

**Comment by hellsan631** [comment]
I wasn't using babel v6 when ```TypeError: (0 , _ava2.default) is not a function``` happened. Still using babel v5, both locally and globally.

**Comment by jamestalmage** [comment]
Oops - the actual error confirmed to be Babel6 is:

```
(0 , _typeof3.default) is not a function.
```

I copy and pasted without noticing the difference.

That said, I suspect you have a babel 6 dependency hanging out.

Assuming you are not on windows, can you run the following **from your project directory**:

```
$ npm ls --depth=0 | grep babel
```

And paste the contents here.

**Comment by hellsan631** [comment]
I don't think so. I am on windows, but I have cygwin installed, so grep still works.

Here are my versions
NPM: 3.3.12
Node: 4.2.2

```
$ npm ls | grep babel
```

Brought up nothing in the local repo. Going for global:

```
$ npm ls -g | grep babel

│ ├─┬ babel-core@5.8.34
│ ├─┬ babel-plugin-espower@1.1.0
│ ├── babel-runtime@5.8.34
│ ├─┬ babelify@6.4.0
```

```babelify@6.4.0``` didn't get babel 6 until ```babelify@7.0.0```

**Comment by jamestalmage** [comment]
Is `ava` only installed globally?

**Comment by hellsan631** [comment]
Yep. It works when its installed only globally. 

When ava is installed __only locally__, the ava command doesn't work.
When ava is installed __only globally__, ava works fine

However in the guide, you recommend one of the first steps being ava --init, which installed ava as a dev-dependency. So ava is installed then both globally and locally. ava then triggers the type error above when I try and run it

**Comment by jamestalmage** [comment]
Try reinstalling locally and running as follows:

`./node_modules/.bin/ava`, that will execute the locally installed version.

Still, the global one should work, as long as it is the same version as your local install.

You may want to also try the following, just to ensure you are not working off a corrupted install:

```
npm cache clean && npm rm -g ava && npm rm ava && npm i -g ava && npm i -D ava
```

**Comment by hellsan631** [comment]
I get various errors when trying to run it locally using the path to bin instead.

```
SyntaxError: Unexpected token case
...
SyntaxError: Unexpected token ILLEGAL
```

Tried the cache clean and still no dice. My full error is here as follows:

```
Uncaught Exception:  C:\Users\HellsAn631\Zend\workspaces\Pippo\test\reader.test.js                            
  TypeError: (0 , _ava2.default) is not a function                                                            
                                                                                                              
Uncaught Exception:  C:\Users\HellsAn631\Zend\workspaces\Pippo\test\reader.test.js                            
  TypeError: (0 , _ava2.default) is not a function                                                            
                                                                                                              
Error: Never got test results from: C:\Users\HellsAn631\Zend\workspaces\Pippo\test\reader.test.js             
    at ChildProcess.<anonymous> (C:\Users\HellsAn631\AppData\Roaming\npm\node_modules\ava\lib\fork.js:61:12)  
    at emitTwo (events.js:87:13)                                                                              
    at ChildProcess.emit (events.js:172:7)                                                                    
    at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)                                 
From previous event:                                                                                          
    at module.exports (C:\Users\HellsAn631\AppData\Roaming\npm\node_modules\ava\lib\fork.js:25:16)            
    at run (C:\Users\HellsAn631\AppData\Roaming\npm\node_modules\ava\cli.js:120:9)                            
    at Array.map (native)                                                                                     
    at C:\Users\HellsAn631\AppData\Roaming\npm\node_modules\ava\cli.js:206:22                                 
From previous event:                                                                                          
    at init (C:\Users\HellsAn631\AppData\Roaming\npm\node_modules\ava\cli.js:198:4)                           
    at Object.<anonymous> (C:\Users\HellsAn631\AppData\Roaming\npm\node_modules\ava\cli.js:243:2)             
    at Module._compile (module.js:435:26)                                                                     
    at Object.Module._extensions..js (module.js:442:10)                                                       
    at Module.load (module.js:356:32)                                                                         
    at Function.Module._load (module.js:311:12)                                                               
    at Function.Module.runMain (module.js:467:10)                                                             
    at startup (node.js:136:18)                                                                               
    at node.js:963:3                                                                                          
```

**Comment by jamestalmage** [comment]
That looks like a stack trace from running globally. What happens with a local run `./node_modules/.bin/ava`

**Comment by hellsan631** [comment]
```
./node_modules/.bin/ava
```
Well, that worked. hmm. Weird, because just a second ago i tried it and got some syntax erros (listed above). but now, after uninstalling it and reinstalling it it works

**Comment by jamestalmage** [comment]
You probably got those because we had you delete `babel` in earlier steps.

If you put the following in package.json:

```json
{
  "scripts" : {
    "test": "ava"
  }
}
```

Then you can run `npm test`, and it will use the local version every time.

**Comment by mnzt** [comment]
Is this to be kept the same, @sindresorhus?

I personally feel a user should be able to run `ava` globally and for the tests to run; it feels much more natural, at least to me.

**Comment by joshuajabbour** [comment]
Really wish the readme clearly stated ava doesn't work with babel 6. I use babel 6 to write my app with es6, but that means it's impossible to use ava to test my app until this issue is resolved. Spent way too long trying to figure this out.

Maybe I'm not understanding things, but this really seems flawed. So if I want to use ava, I'll have to make sure the version of babel used by ava stays in sync with the version I use for my app?

**Comment by mnzt** [comment]
@joshuajabbour when trying to run your tests, are you using `ava foo/bar.js` or are you using `npm test' after initialising Ava?

**Comment by jamestalmage** [comment]
@mnzt - This issue is caused by installing babel 6 locally, not local vs global `ava` cli execution.

FYI:

As of [`0.7.0`](https://github.com/sindresorhus/ava/releases/tag/v0.7.0), the `ava` command will always run the local version of `ava` if available. Even if invoked globally. That release contains a number of breaking changes, so [read the changelog](https://github.com/sindresorhus/ava/releases/tag/v0.7.0) before upgrading.

It should be safe to install `0.7.0` globally, even if your project depends on an earlier version locally. As long as you have run `nmp install` and the older `ava` dependency exists locally it should be used.

```
$ npm install --global ava@latest
```

**Comment by jamestalmage** [comment]
> Really wish the readme clearly stated ava doesn't work with babel 6. 

That is a good point - I will fix that immediately.

> Maybe I'm not understanding things, but this really seems flawed.

I agree. Unfortunately a solution that allows you to use whatever version of Babel you want is not trivial. We are waiting on a few blocking issues in Babel 6, and will transition at that time.

**Comment by joshuajabbour** [comment]
@mnzt Does it matter? Both give me the same error, the one you posted in the original issue.

**Comment by joshuajabbour** [comment]
@mnzt To be more clear, my issue is definitely that I have babel 6 installed and ava doesn't understand how to use it. It sounds like a flaw in the design of ava, because I'm using babel 6 to write my application code. ava should be able to use whatever babel I prefer to parse the test files.

**Comment by sindresorhus** [comment]
@joshuajabbour Ideally, yes, but Babel 6 broke all the things. Babel 6 was also super broken when released with a gadzillion regressions, hence why it has taken us some time to move to it.

**Comment by novemberborn** [comment]
Given the discussion in #292 I'm not sure AVA even needs to use the Babel that's in user's projects. AVA should ship with its own Babel dependency for compiling test files. If users want different behavior they should be able to override test file compilation, which will enable them to use any Babel version they want.

**Comment by sindresorhus** [comment]
@novemberborn Yes, with the Babel 6 change we're not going to use the users Babel, but instead just tell users to use npm@3 if they want to enforce a version as npm will deduplicate it for the user.

**Comment by novemberborn** [comment]
@sindresorhus npm won't be able to dedupe if the user depends on a version outside of the SemVer range of AVA's dependencies. Which is effectively where we are with Babel 5 at the moment. Babel 6 is incompatible and shouldn't be used. In the future a user won't be able to force Babel 7 either.

My understanding of #292 is that AVA provides certain transformations of the test source. If different transformations are required the best way to achieve that is by swapping out AVA's transform process entirely.

If that's the way to go then AVA already shouldn't be pulling in the Babel dependency from the user's project.

**Comment by jamestalmage** [comment]
https://github.com/sindresorhus/ava/pull/221#discussion_r46668343

**Comment by sindresorhus** [comment]
> Which is effectively where we are with Babel 5 at the moment. Babel 6 is incompatible and shouldn't be used. In the future a user won't be able to force Babel 7 either.

That's the whole point. The user would then be able to use Babel 5 themselves, even if we only support Babel 6+, or inverse.

> My understanding of #292 is that AVA provides certain transformations of the test source. If different transformations are required the best way to achieve that is by swapping out AVA's transform process entirely.

Yes, but the above will work for most.

**Comment by jamestalmage** [comment]
The solution discussed in https://github.com/bcoe/nyc/issues/70 could become of use to us in this situation as well.

**Comment by chiefy** [comment]
Wanted to switch from `tape` to `ava`, but I am locked into babel 6 :+1:

**Comment by novemberborn** [comment]
AVA uses Babel 6 now, and if necessary should work with Babel 5 source projects.
